datumaro/components/converter.py

```diff
@@ -34,8 +36,22 @@ def convert(cls, extractor, save_dir, **options):
         return converter.apply()
 
     @classmethod
+    @error_rollback
     def patch(cls, dataset, patch, save_dir, **options):
-        return cls.convert(dataset, save_dir, **options)
+        # Is not any better in performance than just writing a dataset,
+        # but in case of patching (i.e. writing to the previous location),
+        # it allows to avoid many problems with removing and replacing
+        # existing files. Surely, this approach also has problems with
+        # removal of the given directory.
+        tmpdir = save_dir + '.tmp' + uuid.uuid4().hex
```

Comment:
Using `uuid` here seems like an unnecessary indirection (and it also loses a few bits of entropy, since a UUID has some fixed bits). You can just generate the required number of random bytes directly.

Reply:
We need any unused name here. Technically, what `TemporaryDirectory` does, but with normal rights. Maybe we need to copy them from the given directory? And there is also possible the [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) problem.

Reply:
All of what you said makes sense, but it doesn't really address my point, which is that you could've used `os.random(16)` to get more entropy for the same file name length (and without introducing an extra import).

> Technically, what TemporaryDirectory does, but with normal rights. Maybe we need to copy them from the given directory?

It might be a good idea to use `mkdtemp` and copy the permissions from the original directory. That way, you won't need to deal with the randomness by yourself.

> And there is also possible the TOCTOU problem.

TOCTOU can only be a problem if there's a check, and I don't think there is one here. 🙂

Reply:
I tried to re implement this with `mkdtemp`.